python/lsst/daf/butler/core/datasets/type.py

```diff
@@ -472,6 +474,94 @@ def _lookupNames(self) -> Tuple[LookupKey, ...]:
 
         return lookups + self.storageClass._lookupNames()
 
+    def to_simple(self, minimal: bool = False) -> Union[Dict, str]:
+        """Convert this class to a simple python type suitable for
+        serialization.
+
+        Parameters
+        ----------
+        minimal : `bool`, optional
+            Use minimal serialization. Requires Registry to convert
+            back to a full type.
+
+        Returns
+        -------
+        simple : `dict` or `str`
+            The object converted to a dictionary or a simple string.
+        """
+        if minimal:
+            # Only needs the name.
+            return self.name
+
+        # Convert to a dict form
+        as_dict = {"name": self.name,
+                   "storageClass": self._storageClassName,
+                   "isCalibration": self._isCalibration,
+                   "dimensions": self.dimensions.to_simple(),
+                   }
+
+        if self._parentStorageClassName is not None:
+            as_dict["parentStorageClass"] = self._parentStorageClassName
+        return as_dict
+
+    @classmethod
+    def from_simple(cls, simple: Union[Dict, str],
+                    universe: Optional[DimensionUniverse] = None,
+                    registry: Optional[Registry] = None) -> DatasetType:
+        """Construct a new object from the data returned from the `to_simple`
+        method.
+
+        Parameters
+        ----------
+        simple : `dict` of [`str`, `Any`] or `str`
+            The value returned by `to_simple()`.
+        universe : `DimensionUniverse`
+            The special graph of all known dimensions of which this graph will
+            be a subset. Can be `None` if a registry is provided.
+        registry : `lsst.daf.butler.Registry`, optional
+            Registry to use to convert simple name of a DatasetType to
+            a full `DatasetType`. Can be `None` if a full description of
+            the type is provided along with a universe.
+
+        Returns
+        -------
+        datasetType : `DatasetType`
+            Newly-constructed object.
+        """
+        if isinstance(simple, str):
+            if registry is None:
```

Comment:
Same question as with `DatasetRef` as to whether checking for `registry is None` in the outer block would speed things up and/or simplify the code.

Reply:
Is an "is None" test expensive? The instance test presumably costs more (as would a test for a dict with only "name" in it). The client is always going to call this with universe and the server is always going to call this with registry but I'm not sure we can rely on that to be true for all possible JSON serializations.

Reply:
I would guess that `is None` boils down to a pointer comparison in C, so it's definitely going to be faster than `isinstance`, but I'd bet `isinstance(obj, str)` is similarly fast when the object actually is a `str` (I'd bet both cases are dominated by just per-statement interpreter overheads). It's the case where it's going to return `False` that I don't know about; it seems like it _should_ be possible for that to be quite fast, too, for `str` and other builtins, even if it might be much slower for user-defined classes.